operator/v1/types_ingress.go

```diff
@@ -0,0 +1,223 @@
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IngressController describes a managed ingress controller for the cluster. The
+// controller can service OpenShift Route and Kubernetes Ingress resources.
+//
+// When an IngressController is created, a new ingress controller deployment is
+// created to allow external traffic to reach the services that expose Ingress
+// or Route resources. Updating this resource may lead to disruption for public
+// facing network connections as a new ingress controller revision may be rolled
+// out.
+//
+// https://kubernetes.io/docs/concepts/services-networking/ingress-controllers
+//
+// Additional features are available by default and through explicit
+// configuration.
+//
+// 1. To expose the ingress controller to other networks and the Internet, use
+//    the highAvailability configuration.
+//
+// 2. When exposing an ingress controller on supported cloud platforms, managed
+//    wildcard DNS pointing to the ingress controller is automatically enabled.
+//    DNS records are managed only in DNS zones defined in the DNS cluster
+//    configuration resource.
+//
+// 3. If an ingress controller does not specify a default certificate, a new
+//    self-signed certificate valid for the specified domain is generated for
+//    the ingress controller.
+//
+// Whenever possible, sensible defaults for the platform are used. See each
+// field for more details.
+type IngressController struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the IngressController.
+	Spec IngressControllerSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the IngressController.
+	Status IngressControllerStatus `json:"status,omitempty"`
+}
+
+// IngressControllerSpec is the specification of the desired behavior of the
+// IngressController.
+type IngressControllerSpec struct {
+	// domain is a DNS name used to configure various features that help expose
+	// the ingress controller and enable external integrations.
+	//
+	// * The value is published to individual Route statuses so that end-users
+	//   know where to target external DNS records.
+	//
+	// * When wildcard DNS management is enabled, domain is the base domain used
+	//   to construct the wildcard host name.
+	//
+	// * If a generated default certificate is used for the ingress controller,
+	//   the certificate will be valid for domain.
+	//
+	// domain must be unique among all IngressControllers, and cannot be
+	// updated.
+	//
+	// If empty, defaults to the cluster Ingress config domain.
+	//
+	// +optional
+	Domain string `json:"domain,omitempty"`
+
+	// replicas is the desired number of ingress controller replicas. If unset,
+	// defaults to 2.
+	//
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// highAvailability is used to expose the ingress controller and enable high
```

Comment:
This is really platform balancer type or platform balancing strategy.

> What mechanism is used to take traffic from the DNS name to the ingress controller instances?

`Cloud` may be misnamed - is OpenStack a cloud?  I can use service load balancers and program DNS there.  Is RHHI a cloud?  They're going to have some variant of DNS and service load balancers.  

Why are we binding to the host instead of Node Ports in general? Why is NodePorts not an option?

Reply:
I like PlatformBalancingStrategy.

We could expose the underlying Kube concept directly and it would be accurate: Cloud -> LoadBalancerService.

>Why are we binding to the host instead of Node Ports in general? Why is NodePorts not an option?

I assume because the use case involves bypassing the SDN hop into the cluster network. I don't see why NodePorts couldn't be an option but I suspect we still need Host. cc @knobunc

Reply:
Plus existing 3.x topologies are effectively `Host`, so even if perf isn't the concern it seems like we need the option for compatibility

Reply:
Is balancing even accurate in the general sense? PublishingStrategy?

Reply:
From @Miciah on Slack with some other thoughts:
>PointOfPresence
>That's coming from the perspective that this option configures how far we're willing to run the line for you (to a public LB, to the service proxy, or no further than the pod host), but even if that's the right perspective, it's hard to come up with a common-use term that conveys it.
>TerminationPoint, ManagementEndpoint...

Reply:
PointOfPresence is really obscure.

PlatformBalancingStrategy is good.  EndpointPublishingStrategy is good.  Slightly prefer publishing to balancing (this is where you publish the endpoints for the ingress controller).

Reply:
Ran with `EndpointPublishingStrategy`.  PTAL